src/par/map.rs

```diff
@@ -74,13 +78,95 @@ impl<'a, K: Sync, V: Send, S> IntoParallelIterator for &'a mut HashMap<K, V, S>
 impl<K, V, S> FromParallelIterator<(K, V)> for HashMap<K, V, S>
     where K: Eq + Hash + Send,
           V: Send,
-          S: BuildHasher + Default + Send
+          S: BuildHasher + Default + Send + Sync
 {
     fn from_par_iter<P>(par_iter: P) -> Self
         where P: IntoParallelIterator<Item = (K, V)>
     {
-        let mut map = HashMap::default();
-        map.par_extend(par_iter);
+        let mut map = HashMap::<K, V, S>::default();
+        // One shard per core, but round up to a power of two to keep the math simpler.
+        let num_shards = (num_cpus::get() - 1).next_power_of_two();
+        let shard_shift = num_shards.leading_zeros() + 1;
```

Comment:
If `num_cpus::get() == 1`, then `shard_shift` will equal the number of bits in `usize`, which will overflow in `Shr`.

Reply:
Also, `next_power_of_two()` returns a value greater *or equal* -- e.g. 0 -> 1, 1 -> 1, 2 -> 2, 3 -> 4, ... so I don't think you really want to subtract 1 first.